packages/grid/x-data-grid/src/hooks/features/pagination/useGridPagination.ts

```diff
@@ -44,17 +83,214 @@ export const useGridPagination = (
+        mergeStateWithPaginationModel(
+          props.rowCount ?? visibleTopLevelRowCount,
+          props.signature,
+          paginationModel,
+        ),
+        'setPaginationModel',
+      );
+      apiRef.current.forceUpdate();
+    },
+    [apiRef, logger, props.rowCount, props.signature, visibleTopLevelRowCount],
+  );
+
+  const pageApi: GridPaginationApi = {
+    setPage,
+    setPageSize,
+    setPaginationModel,
+  };
+
+  useGridApiMethod(apiRef, pageApi, 'public');
+
+  /**
+   * PRE-PROCESSING
+   */
+  const stateExportPreProcessing = React.useCallback<GridPipeProcessor<'exportState'>>(
+    (prevState, context) => {
+      const paginationModel = gridPaginationModelSelector(apiRef);
+
+      const shouldExportPaginationModel =
+        // Always export if the `exportOnlyDirtyModels` property is not activated
+        !context.exportOnlyDirtyModels ||
+        // Always export if the `paginationModel` is controlled
+        props.paginationModel != null ||
+        // Always export if the `paginationModel` has been initialized
+        props.initialState?.pagination?.paginationModel != null ||
+        // Export if `page` or `pageSize` is not equal to the default value
+        (paginationModel.page !== 0 &&
+          paginationModel.pageSize !== defaultPageSize(props.autoPageSize));
+
+      if (!shouldExportPaginationModel) {
+        return prevState;
+      }
+
+      return {
+        ...prevState,
+        pagination: {
+          ...prevState.pagination,
+          paginationModel,
+        },
+      };
+    },
+    [
+      apiRef,
+      props.paginationModel,
+      props.initialState?.pagination?.paginationModel,
+      props.autoPageSize,
+    ],
+  );
+
+  const stateRestorePreProcessing = React.useCallback<GridPipeProcessor<'restoreState'>>(
+    (params, context) => {
+      const paginationModel = context.stateToRestore.pagination?.paginationModel
+        ? {
+            ...getDefaultGridPaginationModel(props.autoPageSize),
+            ...context.stateToRestore.pagination?.paginationModel,
+          }
+        : gridPaginationModelSelector(apiRef);
+      apiRef.current.updateControlState(
+        'pagination',
+        mergeStateWithPaginationModel(
+          props.rowCount ?? visibleTopLevelRowCount,
+          props.signature,
+          paginationModel,
+        ),
+        'stateRestorePreProcessing',
+      );
+      return params;
+    },
+    [apiRef, props.autoPageSize, props.rowCount, props.signature, visibleTopLevelRowCount],
+  );
+
+  useGridRegisterPipeProcessor(apiRef, 'exportState', stateExportPreProcessing);
+  useGridRegisterPipeProcessor(apiRef, 'restoreState', stateRestorePreProcessing);
+
+  /**
+   * EVENTS
+   */
+  const handlePaginationModelChange: GridEventListener<'paginationModelChange'> = () => {
+    const paginationModel = gridPaginationModelSelector(apiRef);
+    if (apiRef.current.virtualScrollerRef?.current) {
+      apiRef.current.scrollToIndexes({
```

Comment:
This changes the behavior a bit - previously the page was scrolled to the top only on page change. Currently, it would scroll to the top even if the page size changes

Reply:
Yes and it seems to make sense to me too. I mean after `pageSize` change, the visible rows are computed again and the same row isn't necessarily shown on the viewport. Is there a catch you can observe here?